Rework main and add graceful shutdown signal

LucioFranco opened the pull request:
This change adds a shutdown signal and creates its own tokio runtime
to be able to handle the shutdown. This implementation is loosely inspired by how linkerd2-proxy handles shutdowns and the tokio runtime.

hntd187 commented:
Also this doesn't look like it accounts for any signal beyond control-c? What about the differences in windows and unix signals?

LucioFranco commented:
For reference this is how linkerd2-proxy handles the differences between windows and unix.

https://github.com/linkerd/linkerd2-proxy/blob/master/src/signal.rs

fhaynes commented on Cargo.toml:
```diff
@@ -18,7 +18,8 @@ path = "src/lib.rs"
 [dependencies]
-gotham            = "^0.3"
+#gotham            = "^0.3"
+gotham            = { git = "https://github.com/LucioFranco/gotham" }
```
I thought we couldn't use git deps in prod?

LucioFranco replied:
Now it points to a specific commit on `gotham-rs/gotham`

fhaynes commented on Cargo.toml:
```diff
@@ -28,6 +29,7 @@ serde_json        = "^1.0"
 futures           = "^0.1"
 tantivy           = "^0.7"
 tokio             = "^0.1"
+tokio-signal 	  = "^0.2"
```
Line up the `=` please

fhaynes commented on src/bin/main.rs:
```diff
@@ -1,40 +1,98 @@
 
-use clap::{App, Arg, ArgMatches};
+pub fn main() -> Result<(), ()> {
+    // Get the toshi settings
```
This comment doesn't really add anything, I'd suggest removing it.

fhaynes commented on src/bin/main.rs:
```diff
@@ -1,40 +1,98 @@
+        server.select(shutdown)
+    };
+
+    rt.spawn(toshi.map(|_| ()).map_err(|_| ()));
```
Do we want to throw away the error(s) here?

LucioFranco replied:
Totally agree, how do you think we should we just print it as an error or gracefully shutdown?

fhaynes replied:
At a minimum we should print it. If it is something we can't recover from, gracefully shut down.

LucioFranco replied:
Actually just realized we can throw away the errors here since, toshi is a select future. The future that created the select is already a `Future<Item = (), Error = ()>` and all a select does is augment it with a return type of `Future<Item = ((), ()), Error = ((), ())>`. So this is fine as there are no actual errors passed. The map is there to fit those return types to the one required by `spawn`.

hntd187 replied:
So should we go up that chain and propagate real errors down?

fhaynes commented on src/bin/main.rs:
```diff
@@ -1,40 +1,98 @@
+    // Gracefully shutdown the tokio runtime when a shutdown message has been
+    // received on the shutdown_signal channel.
+    shutdown_signal
+        .map_err(|_| ())
```
Same ? as above

fhaynes commented on src/bin/main.rs:
```diff
@@ -102,67 +160,80 @@ pub fn runner() -> i32 {
+    if settings.enable_clustering {
+        // Run the tokio runtime, this will start an event loop that will process
+        // the connect_consul future. It will block until the future is completed
+        // by either completing successfuly or erroring out.
```
```suggestion
        // by either completing successfully or erroring out.
```

hntd187 commented on src/commit.rs:
```diff
@@ -1,36 +1,26 @@
-        let catalog = Arc::clone(&self.catalog);
-        let task = Interval::new(Instant::now(), Duration::from_secs(self.commit_duration))
+    pub fn start(self) {
+        let catalog = self.catalog.clone();
```
For cloning an `Arc` you want to use `Arc::clone` and not `.clone()` https://doc.rust-lang.org/std/sync/struct.Arc.html#cloning-references

LucioFranco replied:
👍 Totally did not know this was a thing and that `Arc::clone` was idiomatic. Will update the code to use this. Thanks.

LucioFranco commented:
@hntd187 yeah, was planning on doing it this weekend.

LucioFranco commented:
I've updated this PR and rebased against master. I left out the changes to using other signals, as I would rather open an issue and deal with that in another PR. This should be ready for merge.
